File: citadel-first/backend/app/schemas/trust_dividend.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, Field


class DividendPaymentStatus(str, Enum):
    PENDING = "PENDING"
    PAID = "PAID"


# ── Dividend Create (admin/Vanguard) ────────────────────────────

class TrustDividendCreateRequest(BaseModel):
    trust_portfolio_id: int
    dividend_amount: Decimal = Field(gt=0, max_digits=15, decimal_places=2)
    trustee_fee_amount: Decimal = Field(default=Decimal("0"), ge=0, max_digits=15, decimal_places=2)
    period_starting_date: date | None = None
    period_ending_date: date | None = None
    dividend_quarter: int = Field(default=0, ge=0, le=4)


# ── Dividend Status Update ──────────────────────────────────────

class TrustDividendStatusUpdateRequest(BaseModel):
    payment_status: DividendPaymentStatus
    payment_date: date | None = None


# ── Dividend Response ───────────────────────────────────────────

class TrustDividendResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    trust_portfolio_id: int
    reference_number: str
    dividend_amount: Decimal
    trustee_fee_amount: Decimal
    period_starting_date: date | None
    period_ending_date: date | None
    dividend_quarter: int
    payment_status: str
    payment_date: date | None
    created_at: datetime | None
    updated_at: datetime | None


class TrustDividendListResponse(BaseModel):
    dividends: list[TrustDividendResponse]

File: citadel-first/mobile/lib/models/trust_dividend.dart
import 'package:equatable/equatable.dart';

class TrustDividend extends Equatable {
  final int id;
  final int trustPortfolioId;
  final String referenceNumber;
  final double dividendAmount;
  final double trusteeFeeAmount;
  final DateTime? periodStartingDate;
  final DateTime? periodEndingDate;
  final int dividendQuarter;
  final String paymentStatus;
  final DateTime? paymentDate;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const TrustDividend({
    required this.id,
    required this.trustPortfolioId,
    required this.referenceNumber,
    required this.dividendAmount,
    required this.trusteeFeeAmount,
    this.periodStartingDate,
    this.periodEndingDate,
    this.dividendQuarter = 0,
    this.paymentStatus = 'PENDING',
    this.paymentDate,
    this.createdAt,
    this.updatedAt,
  });

  factory TrustDividend.fromJson(Map<String, dynamic> json) {
    return TrustDividend(
      id: json['id'] as int,
      trustPortfolioId: json['trust_portfolio_id'] as int,
      referenceNumber: json['reference_number'] as String,
      dividendAmount: _parseDouble(json['dividend_amount']) ?? 0.0,
      trusteeFeeAmount: _parseDouble(json['trustee_fee_amount']) ?? 0.0,
      periodStartingDate: json['period_starting_date'] != null
          ? DateTime.tryParse(json['period_starting_date'] as String)
          : null,
      periodEndingDate: json['period_ending_date'] != null
          ? DateTime.tryParse(json['period_ending_date'] as String)
          : null,
      dividendQuarter: json['dividend_quarter'] as int? ?? 0,
      paymentStatus: json['payment_status'] as String? ?? 'PENDING',
      paymentDate: json['payment_date'] != null
          ? DateTime.tryParse(json['payment_date'] as String)
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.tryParse(json['updated_at'] as String)
          : null,
    );
  }

  String get displayAmount {
    return 'RM ${dividendAmount.toStringAsFixed(2).replaceAll(RegExp(r'\B(?=(\d{3})+(?!\d))'), ',')}';
  }

  String get displayTrusteeFee {
    return 'RM ${trusteeFeeAmount.toStringAsFixed(2).replaceAll(RegExp(r'\B(?=(\d{3})+(?!\d))'), ',')}';
  }

  String get netAmount {
    final net = dividendAmount - trusteeFeeAmount;
    return 'RM ${net.toStringAsFixed(2).replaceAll(RegExp(r'\B(?=(\d{3})+(?!\d))'), ',')}';
  }

  String get quarterLabel => dividendQuarter > 0 ? 'Q$dividendQuarter' : '';

  String get paymentStatusLabel => switch (paymentStatus) {
        'PENDING' => 'Pending',
        'PAID' => 'Disbursed',
        _ => paymentStatus,
      };

  String displayDate(DateTime? date) {
    if (date == null) return 'N/A';
    return '${date.day.toString().padLeft(2, '0')}/${date.month.toString().padLeft(2, '0')}/${date.year}';
  }

  @override
  List<Object?> get props => [id, referenceNumber];
}

double? _parseDouble(dynamic value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value);
  return null;
}